src/CoreWCF.Primitives/src/CoreWCF/Security/LdapAdapter.cs

```diff
@@ -0,0 +1,119 @@
+            var retrievedClaims = new List<string>();
+            if (!string.IsNullOrEmpty(settings.OrgUnit))
+            {
+                distinguishedName = "OU=" + settings.OrgUnit + "," + distinguishedName;
+            }
+
+            var genericFilter = $"(&(objectClass=user)(sAMAccountName={userAccountName}))"; // This is using ldap search query language, it is looking on the server for someUser
+            var upnFilter = $"(&(objectClass=user)(userPrincipalName={originalUserName}))";
+            var genericSearchRequest = new SearchRequest(distinguishedName, genericFilter, SearchScope.Subtree, null);
+            var upnSearchRequest = new SearchRequest(distinguishedName, upnFilter, SearchScope.Subtree, null);
+            SearchResponse searchResponse = null;
+            try
+            {
+                if (upnIndex > 0)
+                {
+                    searchResponse = (SearchResponse)await Task<DirectoryResponse>.Factory.FromAsync(
+                    settings.LdapConnection.BeginSendRequest, settings.LdapConnection.EndSendRequest,
+                    upnSearchRequest, PartialResultProcessing.NoPartialResultSupport, null);
+                    if(searchResponse !=null && searchResponse.Entries != null && searchResponse.Entries.Count > 1)
+                    {
+                        throw new Exception(SR.DuplicateUPN); //resource
+                    }
+                }
+                if (searchResponse == null || searchResponse.Entries == null || searchResponse.Entries.Count == 0)
+                {
+                    searchResponse = (SearchResponse)await Task<DirectoryResponse>.Factory.FromAsync(
+                    settings.LdapConnection.BeginSendRequest, settings.LdapConnection.EndSendRequest,
+                    genericSearchRequest, PartialResultProcessing.NoPartialResultSupport, null);
+                }
+            }
+            catch (Exception ex)
+            {
+
+                if (searchResponse?.ErrorMessage != null)
+                {
+                    throw new Exception(searchResponse.ErrorMessage);
+                }
+                else
+                {
+                    throw ex;
+                }
+            }
+
+            if (searchResponse.Entries.Count > 0)
+            {
+                var userFound = searchResponse.Entries[0]; //Get the object that was found on ldap
+                var memberof = userFound.Attributes["memberof"]; // You can access ldap Attributes with Attributes property
+
+                foreach (var group in memberof)
+                {
+                    // Example distinguished name: CN=TestGroup,DC=KERB,DC=local
+                    var groupDN = $"{Encoding.UTF8.GetString((byte[])group)}";
+                    var groupCN = groupDN.Split(',')[0].Substring("CN=".Length);
+                    retrievedClaims.Add(groupCN);
```

Comment:
There should be a bit more validation here. Make sure group isn't null, make sure groupDn isn't the empty string. Make sure the first entry does begin with `CN=`. Also I believe you can have leading or trailing spaces in many places, so `CN = TestGroup, DC = KERB, DC = local` is technically a valid distinguished name. You need to split the groupDN on a comma, then split the first field on the equals sign and take the second half (after checking it split into exactly 2 pieces) and Trim() it to remove and leading or trailing spaces.